### ISSUE TITLE
Glibc not found error on Frontera with Miniconda installation

### ISSUE THREAD

**Issue by kks32** [geoelements/gns#3]
Including `torch_geometric` on TACC Frontera node results in:

```python
from torch_geometric.nn import MessagePassing, radius_graph
```

```shell
Traceback (most recent call last):
  File "gns.py", line 3, in <module>
    from torch_geometric.nn import MessagePassing, radius_graph
  File "/work/05873/kks32/frontera/miniconda3/envs/gns/lib/python3.8/site-packages/torch_geometric/nn/__init__.py", line 3, in <module>
    from .sequential import Sequential
  File "/work/05873/kks32/frontera/miniconda3/envs/gns/lib/python3.8/site-packages/torch_geometric/nn/sequential.py", line 10, in <module>
    from torch_geometric.nn.conv.utils.jit import class_from_module_repr
  File "/work/05873/kks32/frontera/miniconda3/envs/gns/lib/python3.8/site-packages/torch_geometric/nn/conv/__init__.py", line 24, in <module>
    from .spline_conv import SplineConv
  File "/work/05873/kks32/frontera/miniconda3/envs/gns/lib/python3.8/site-packages/torch_geometric/nn/conv/spline_conv.py", line 16, in <module>
    from torch_spline_conv import spline_basis, spline_weighting
  File "/work/05873/kks32/frontera/miniconda3/envs/gns/lib/python3.8/site-packages/torch_spline_conv/__init__.py", line 15, in <module>
    torch.ops.load_library(spec.origin)
  File "/work/05873/kks32/frontera/miniconda3/envs/gns/lib/python3.8/site-packages/torch/_ops.py", line 110, in load_library
    ctypes.CDLL(path)
  File "/work/05873/kks32/frontera/miniconda3/envs/gns/lib/python3.8/ctypes/__init__.py", line 373, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by /work2/05873/kks32/frontera/miniconda3/envs/gns/lib/python3.8/site-packages/torch_spline_conv/_basis_cuda.so)
```
Doing an `ldd /work2/05873/kks32/frontera/miniconda3/envs/gns/lib/python3.8/site-packages/torch_spline_conv/_basis_cuda.so` yields:

```
/work2/05873/kks32/frontera/miniconda3/envs/gns/lib/python3.8/site-packages/torch_spline_conv/_basis_cuda.so: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by /work2/05873/kks32/frontera/miniconda3/envs/gns/lib/python3.8/site-packages/torch_spline_conv/_basis_cuda.so)
	linux-vdso.so.1 =>  (0x00007fff9cfc0000)
	/opt/apps/xalt/xalt/lib64/libxalt_init.so (0x00002af9273c2000)
	libc10.so => not found
	libtorch.so => not found
	libtorch_cpu.so => not found
	libtorch_python.so => not found
	libcudart.so.11.0 => /work2/05873/kks32/frontera/miniconda3/envs/gns/lib/python3.8/site-packages/torch_spline_conv/../../../libcudart.so.11.0 (0x00002af927605000)
	libc10_cuda.so => not found
	libtorch_cuda_cu.so => not found
	libtorch_cuda_cpp.so => not found
	libstdc++.so.6 => /work2/05873/kks32/frontera/miniconda3/envs/gns/lib/python3.8/site-packages/torch_spline_conv/../../../libstdc++.so.6 (0x00002af9278a2000)
	libm.so.6 => /lib64/libm.so.6 (0x00002af927a17000)
	libgcc_s.so.1 => /work2/05873/kks32/frontera/miniconda3/envs/gns/lib/python3.8/site-packages/torch_spline_conv/../../../libgcc_s.so.1 (0x00002af927329000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00002af927d19000)
	libc.so.6 => /lib64/libc.so.6 (0x00002af927f35000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00002af928303000)
	/lib64/ld-linux-x86-64.so.2 (0x00002af92719e000)
	librt.so.1 => /lib64/librt.so.1 (0x00002af928507000)
```

**Comment by jpvantassel** [comment]
Appears to be an incompatibility between the virtual environment and real environment. To solve the issue I moved us to Python 3.9 and did the module loads before conda environment initialization.

I created a `make_env.sh` script for reproducibility, see below. Will add to dev branch as well.

```
#!/bin/bash

# Assumes you already have conda installed
# If not run the following:
#   wget https://repo.anaconda.con/miniconda/Miniconda3-latest-Linux-x86_64.sh
#   bash Miniconda3-latest-Linux-x64_64.sh
#   bash

conda env remove --name gnn

module reset
module load python3/3.9.2
module load cuda/11.0 cudnn/8.0.5

conda create --name gnn python=3.9.2
conda activate gnn
conda install pytorch torchvision torchaudio cudatoolkit=11.0 -c pytorch
conda install pyg -c pyg -c conda-forge
```

**Comment by kks32** [comment]
FYI: TensorFlow does not yet support Python3.10

**Comment by jpvantassel** [comment]
Two main issues here:
1) torch_spline_conv was built using Ubuntu so their .whl are invalid for running on frontera, fixed by switching to pip and rebuilding to avoid .whl.
2) invalid cuda/pytorch/pyg combination, fixed by building cuda 11.1 on frontera. Code below for creating env.

```
#!/bin/bash

module reset # -> default is python3/3.7.0

# background
# ----------

# pytorch
# pytorch 1.10 + cuda -> requires cuda 10.2 or 11.3
# pytorch 1.9 + cuda -> (not listed for some reason?)
#                       based on pygeometric it supports cuda 10.2 or 11.1
# pytorch 1.8 + cuda -> requires cuda 10.2 or 11.1

# pygeometric 
# pygeometric -> requires pytorch >= 1.8 

# frontera
# only has cuda 10 10.1 11.0 11.3 


# create env
# ---------
ml use /work2/02604/ajs2987/frontera/apps/modulefiles
ml cuda/11.1
module load python3/3.9.2

python3 -m virtualenv venv
source venv/bin/activate

which pip3
pip3 install torch==1.10.0+cu111 torchvision==0.11.1+cu111 torchaudio==0.10.0+cu111 -f https://download.pytorch.org/whl/cu111/torch_stable.html
pip3 install torch-spline-conv -f https://data.pyg.org/whl/torch-1.10.0+cu111.html --no-binary torch-spline-conv
pip3 install torch-scatter torch-sparse torch-cluster torch-geometric -f https://data.pyg.org/whl/torch-1.10.0+cu111.html

# test env
# --------
which python
echo 'test_pytorch.py'
python test_pytorch.py

echo 'test_pytorch_cuda_gpu.py'
python test_pytorch_cuda_gpu.py

echo 'test_torch_geometric.py'
#python test_torch_geometric.py


# Clean up
# --------
#deactivate
#rm -r venv```
